src/NUnitFramework/framework/Constraints/AnyOfConstraint.cs

```diff
@@ -0,0 +1,172 @@
+// ***********************************************************************
+// Copyright (c) 2007 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace NUnit.Framework.Constraints
+{
+    /// <summary>
+    /// <see cref="AnyOfConstraint"/> is used to determine whether item is equal to any of parameters
+    /// </summary>
+    public class AnyOfConstraint : Constraint
+    {
+        private readonly IEnumerable _expected;
+
+        /// <summary>
+        /// The NUnitEqualityComparer in use for this constraint
+        /// </summary>
+        private readonly NUnitEqualityComparer _comparer = NUnitEqualityComparer.Default;
+
+        /// <summary>
+        /// Construct a OneOfConstraint
+        /// </summary>
+        /// <param name="expected">Expected collection</param>
+        public AnyOfConstraint(IEnumerable expected)
+        {
+            _expected = expected;
+        }
+
+        /// <summary>
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
+        /// </summary>
+        public override string Description
+        {
+            get
+            {
+                return "any of " + MsgUtils.FormatValue(_expected);
```

Comment:
I had to do quite a bit of code searching to figure out what the description would look like and particularly whether it would be truncated. It appears that it will be in the form `any of <1, 42, 99, 1234, 4321>` and no truncation will occur.

I'm thinking we should
1. Decide on the format. You might think that array formats are already decided by `MsgUtils` but in this case we use a `params` array to model separate arguments. Maybe we want `1 OR 42 OR 99` or something equally silly.
2. Add a little comment here if we keep the present format so folks won't have to search.
3. Add tests of the description.
4. Add tests of the error message.

Reply:
`any of <2, 3, 5>` seems to be most consistent by the rest of NUnit. :+1:

Reply:
Speaking as the devil's advocate... It's the standard representation for collections. Here, a user can use the syntax without ever thinking about arrays or collections. In some ways it's like the Between syntax.

Another weird idea... what if the argument were a tuple?

Reply:
I like the similarity to collections and do see `EqualTo.Or.EqualTo...` as semantically identical to talking about a collection.
Anyway, the user will be typing values separated by comma and space, so let's keep at least that aspect. The prior syntax `<or <equal 2> <or <equal 3> <equal 5>>>` might suggest that what we really want is `<any of 2, 3, 5>`?

Tuple... interesting. I don't know?

Reply:
@CharliePoole what would you suggest as ToString result? Currently all the collection constraints (it is still a collection constraint underneath) show  <.. Type[]>, not actual values

Reply:
Actually, it's not a `CollectionConstraint` in the sense I meant. It inherits from `Constraint`. Our `CollectionConstraints` apply to actual values that are collections. 

The only other multiple argument constraint we have is `RangeConstraint`, which gives you something like `<range 5 42>`. While we could live with `<oneof Type[]>` I'd rather see the list because it would make for better testing.